File: scripts/build_catalog.py
#!/usr/bin/env python3
"""
Gera o catálogo estático de labs a partir de labs/*/lab.yaml.

Uso:
    pip install pyyaml jsonschema
    python3 scripts/build_catalog.py

Escreve:
    site/public/catalog.json       — índice leve de todos os labs válidos
    site/public/labs/<id>.json     — lab completo (equivalente ao lab.yaml,
                                      em JSON) para consumo pela CLI

Só escreve alguma coisa se TODOS os labs em /labs forem válidos — um
catálogo parcial/quebrado é pior do que nenhum catálogo. Se algum lab
falhar a validação, roda `python3 scripts/validate_lab_schema.py` para
ver o erro detalhado.

Exit code 0 em sucesso, 1 se algum lab for inválido.
"""
import json
import sys
from datetime import datetime, timezone
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent))

from _lab_common import LABS_DIR, iter_lab_results  # noqa: E402

ROOT = Path(__file__).resolve().parent.parent
SITE_PUBLIC_DIR = ROOT / "site" / "public"
CATALOG_PATH = SITE_PUBLIC_DIR / "catalog.json"
LABS_JSON_DIR = SITE_PUBLIC_DIR / "labs"

SCHEMA_VERSION = 1

# Imagens ainda não construídas/publicadas pela equipe (ver
# docs/MAPA_DO_REPO.md e docs/CHECKLIST.md Fase 2). Heurística por padrão
# de nome: "SEU_USUARIO" é sempre placeholder; "ghcr.io/codisec/lab-*" é
# onde as imagens reais vão morar quando publicadas, mas nenhuma foi
# publicada ainda — quando isso mudar, revisar esta heurística (ex: manter
# uma lista explícita de ids já publicados em vez de inferir pela URL).
_PLACEHOLDER_MARKERS = ("SEU_USUARIO", "ghcr.io/codisec/lab-")


def is_image_published(image: str) -> bool:
    return not any(marker in image for marker in _PLACEHOLDER_MARKERS)


def catalog_entry(lab: dict) -> dict:
    return {
        "id": lab["id"],
        "title": lab["title"],
        "category": lab["category"],
        "difficulty": lab["difficulty"],
        "duration": lab["duration"],
        "description": lab["description"],
        "tags": lab.get("tags", []),
        "image": lab["image"],
        "image_published": is_image_published(lab["image"]),
        "task_count": len(lab["tasks"]),
        "maintainers": lab.get("maintainers", []),
    }


def main() -> int:
    results = list(iter_lab_results())
    invalid = [r for r in results if not r.valid and not r.skipped]
    skipped = [r for r in results if r.skipped]

    if invalid or skipped or not results:
        print("❌ Catálogo não gerado — nem todos os labs são válidos.")
        print("   Rode `python3 scripts/validate_lab_schema.py` para ver o detalhe.")
        return 1

    labs = [dict(r.data, _dir=r.dir_name) for r in results]
    labs.sort(key=lambda lab: lab["id"])

    SITE_PUBLIC_DIR.mkdir(parents=True, exist_ok=True)
    LABS_JSON_DIR.mkdir(parents=True, exist_ok=True)

    catalog = {
        "schema_version": SCHEMA_VERSION,
        "generated_at": datetime.now(timezone.utc).isoformat(timespec="seconds"),
        "total": len(labs),
        "labs": [catalog_entry(lab) for lab in labs],
    }
    CATALOG_PATH.write_text(
        json.dumps(catalog, ensure_ascii=False, indent=2) + "\n", encoding="utf-8"
    )

    for lab in labs:
        lab_out = {k: v for k, v in lab.items() if k != "_dir"}
        lab_out["image_published"] = is_image_published(lab["image"])
        (LABS_JSON_DIR / f"{lab['id']}.json").write_text(
            json.dumps(lab_out, ensure_ascii=False, indent=2) + "\n", encoding="utf-8"
        )

    print(f"✅ Catálogo gerado: {len(labs)} labs.")
    print(f"   {CATALOG_PATH.relative_to(ROOT)}")
    print(f"   {LABS_JSON_DIR.relative_to(ROOT)}/<id>.json ({len(labs)} arquivos)")
    unpublished = [lab["id"] for lab in labs if not is_image_published(lab["image"])]
    if unpublished:
        print(
            f"   ⚠️  {len(unpublished)} lab(s) com imagem Docker ainda não "
            f"publicada: {', '.join(unpublished)}"
        )
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: site/scripts/build-catalog.mjs
#!/usr/bin/env node
// Gera o catálogo estático de labs a partir de labs/*/lab.yaml.
//
// Substitui a versão antiga em Python (scripts/build_catalog.py):
// aquela dependia de `pip install pyyaml jsonschema` rodar no
// ambiente de build, o que funcionava localmente (instalado à mão)
// mas quebrou o deploy na Cloudflare Pages, que não tem Python
// configurado por padrão no pipeline do site. Esta versão é 100% Node
// — instalada junto com o resto via `npm ci`, sem dependência cruzada
// de runtime no build.
//
// labs/schema.json continua sendo a fonte de verdade única. Ela é
// validada aqui via ajv (JSON Schema Draft-07) e, separadamente, por
// scripts/validate_lab_schema.py em Python (usado só por
// .github/workflows/validate-labs.yml, fora do pipeline de deploy do
// site) — duas implementações independentes e maduras da MESMA spec
// declarativa, não duas cópias de lógica de validação escrita à mão.
//
// Uso: node scripts/build-catalog.mjs (chamado por `npm run catalog`,
// que por sua vez roda automaticamente antes de `npm run dev`/`build`)
//
// Escreve public/catalog.json e public/labs/<id>.json. Só escreve
// alguma coisa se TODOS os labs forem válidos — um catálogo
// parcial/quebrado é pior que nenhum catálogo.

import { mkdirSync, readdirSync, readFileSync, statSync, writeFileSync } from 'node:fs';
import { dirname, join } from 'node:path';
import { fileURLToPath } from 'node:url';
import Ajv from 'ajv';
import { load as loadYaml } from 'js-yaml';

const __dirname = dirname(fileURLToPath(import.meta.url));
const SITE_DIR = join(__dirname, '..');
const ROOT = join(SITE_DIR, '..');
const LABS_DIR = join(ROOT, 'labs');
const SCHEMA_PATH = join(LABS_DIR, 'schema.json');
const PUBLIC_DIR = join(SITE_DIR, 'public');
const LABS_JSON_DIR = join(PUBLIC_DIR, 'labs');
const CATALOG_PATH = join(PUBLIC_DIR, 'catalog.json');

const SCHEMA_VERSION = 1;

// Imagens ainda não construídas/publicadas pela equipe (ver
// docs/MAPA_DO_REPO.md e docs/CHECKLIST.md Fase 2). Heurística por
// padrão de nome — se mudar aqui, mudar também no comentário
// equivalente em site/src/pages/labs/[id].astro.
const PLACEHOLDER_MARKERS = ['SEU_USUARIO', 'ghcr.io/codisec/lab-'];

function isImagePublished(image) {
  return !PLACEHOLDER_MARKERS.some((marker) => image.includes(marker));
}

function loadLabResults() {
  const schema = JSON.parse(readFileSync(SCHEMA_PATH, 'utf-8'));
  const ajv = new Ajv({ allErrors: true, strict: false });
  const validate = ajv.compile(schema);

  const dirNames = readdirSync(LABS_DIR)
    .filter((name) => statSync(join(LABS_DIR, name)).isDirectory())
    .sort();

  return dirNames.map((dirName) => {
    const labFile = join(LABS_DIR, dirName, 'lab.yaml');

    let raw;
    try {
      raw = readFileSync(labFile, 'utf-8');
    } catch (err) {
      if (err.code === 'ENOENT') return { dirName, skipped: true };
      throw err;
    }

    let data;
    try {
      data = loadYaml(raw);
    } catch (err) {
      return { dirName, yamlError: String(err.message ?? err) };
    }

    if (!validate(data)) {
      const messages = validate.errors.map(
        (e) => `[${e.instancePath || '(raiz)'}] ${e.message}`
      );
      return { dirName, data, schemaErrors: messages };
    }

    if (data.id !== dirName) {
      return {
        dirName,
        data,
        idMismatch: `campo 'id' (${data.id}) precisa ser igual ao nome da pasta (${dirName}).`,
      };
    }

    return { dirName, data };
  });
}

function isValidResult(r) {
  return !r.skipped && r.data != null && !r.yamlError && !r.schemaErrors && !r.idMismatch;
}

function catalogEntry(lab) {
  return {
    id: lab.id,
    title: lab.title,
    category: lab.category,
    difficulty: lab.difficulty,
    duration: lab.duration,
    description: lab.description,
    tags: lab.tags ?? [],
    image: lab.image,
    image_published: isImagePublished(lab.image),
    requires_privileged: lab.requires_privileged ?? false,
    required_capabilities: lab.required_capabilities ?? [],
    task_count: lab.tasks.length,
    maintainers: lab.maintainers ?? [],
  };
}

function main() {
  const results = loadLabResults();
  const invalid = results.filter((r) => !r.skipped && !isValidResult(r));
  const skipped = results.filter((r) => r.skipped);

  if (invalid.length > 0 || skipped.length > 0 || results.length === 0) {
    console.error('❌ Catálogo não gerado — nem todos os labs são válidos.');
    console.error('   Rode `python3 scripts/validate_lab_schema.py` para ver o detalhe.');
    process.exit(1);
  }

  const labs = results.map((r) => r.data).sort((a, b) => a.id.localeCompare(b.id));

  mkdirSync(PUBLIC_DIR, { recursive: true });
  mkdirSync(LABS_JSON_DIR, { recursive: true });

  const catalog = {
    schema_version: SCHEMA_VERSION,
    generated_at: new Date().toISOString(),
    total: labs.length,
    labs: labs.map(catalogEntry),
  };
  writeFileSync(CATALOG_PATH, JSON.stringify(catalog, null, 2) + '\n', 'utf-8');

  for (const lab of labs) {
    const labOut = {
      ...lab,
      image_published: isImagePublished(lab.image),
      requires_privileged: lab.requires_privileged ?? false,
      required_capabilities: lab.required_capabilities ?? [],
    };
    writeFileSync(
      join(LABS_JSON_DIR, `${lab.id}.json`),
      JSON.stringify(labOut, null, 2) + '\n',
      'utf-8'
    );
  }

  console.log(`✅ Catálogo gerado: ${labs.length} labs.`);
  console.log('   public/catalog.json');
  console.log(`   public/labs/<id>.json (${labs.length} arquivos)`);

  const unpublished = labs.filter((l) => !isImagePublished(l.image)).map((l) => l.id);
  if (unpublished.length > 0) {
    console.log(
      `   ⚠️  ${unpublished.length} lab(s) com imagem Docker ainda não publicada: ${unpublished.join(', ')}`
    );
  }
}

main();
